feat(lsp): add related_information to vim.Diagnostic

tom-anders opened the pull request:
Basically a simplified version of https://github.com/neovim/neovim/issues/19649#issuecomment-1750272564

Looking to get some initial feedback here. Instead of just appending to the message, we might instead want to store the related info in the `vim.Diagnostic` struct, and then instead extend the logic in `vim.diagnostic.open_float()`...?

This PR:
![Screenshot_2024-05-04_21-20-17](https://github.com/neovim/neovim/assets/13141438/bde8cbc2-fbf4-4264-9f5e-fb58a37f2d87)
vscode:
![Screenshot_2024-05-04_21-22-00](https://github.com/neovim/neovim/assets/13141438/efc56625-284b-4242-8a84-d824fa035dc9)

EDIT: PR now only adds the related_information field to the toplevel of vim.Diagnostic (see discussion below). We could then decide in a follow-up PR if and how we should display this to the user.

wookayin commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -110,6 +110,14 @@ local function diagnostic_lsp_to_vim(diagnostics, bufnr, client_id)
   return vim.tbl_map(function(diagnostic)
     local start = diagnostic.range.start
     local _end = diagnostic.range['end']
+    for _, relatedInfo in pairs(diagnostic.relatedInformation or {}) do
```
`relatedInformation` is an array, so it looks like `ipairs` instead of `pairs` would be better?

wookayin commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -110,6 +110,14 @@ local function diagnostic_lsp_to_vim(diagnostics, bufnr, client_id)
     local start = diagnostic.range.start
     local _end = diagnostic.range['end']
+    for _, relatedInfo in pairs(diagnostic.relatedInformation or {}) do
+      if relatedInfo.location then
```
`location` seems to be a non-nil (required) field of the `lsp.DiagnosticRelatedInformation` class.

mfussenegger commented:
I think instead of extending the message this should go into the structure, otherwise users can't customise this. (And I'd wager there's a bunch who don't want this in the message).

There are also actions where we send the diagnostic back to the server and I think for the server to be able to re-associate the payloads with their internal state the diagnostic mustn't change.

tom-anders committed:
feat(lsp): add vim.diagnostic.related_information field

For now this just adds the field to the toplevel Diagnostic structure
and fills it with the corresponding struct from LSP.
We can decide in a follow-up if and how we should display these to the
user.

tom-anders commented:
> I think instead of extending the message this should go into the structure, otherwise users can't customise this. (And I'd wager there's a bunch who don't want this in the message).
> 
> There are also actions where we send the diagnostic back to the server and I think for the server to be able to re-associate the payloads with their internal state the diagnostic mustn't change.

Sounds fair - I changed the PR so that now it only adds the `relatedInformation` field to the toplevel of `vim.Diagnostic`. We could then decide in a follow-up PR if and how we should display this to the user.

If this looks good, I'll add a test to `diagnostic_spec.lua` that covers this new field.

tom-anders commented:
> `vim.Diagnostic` is not an "LSP diagnostic" though, and we try very hard to avoid imposing LSP specific concepts into the (generic) diagnostic structure. In this case it is probably fine since the "related information" seems pretty generic (could be used by non-LSP diagnostic producers).
> 
> If this does need to be in the top level structure it should follow the standard naming conventions (`related_information`, not `relatedInformation`). The latter style indicates that it is a field that maps directly to some LSP concept.

Yeah to me this seems general enough (I could imagine something like overseer.nvim make use of this). Thanks for the hint related to naming conventions, fixed!

gpanders commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -42,6 +45,29 @@ local M = {}
 
+--- Related message and location for a diagnostic.
+--- This can be used to point to code locations that cause or are related to
+--- a diagnostics, e.g when duplicating a symbol in a scope.
```
```suggestion
--- a diagnostic, e.g when duplicating a symbol in a scope.
```

gpanders commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -97,6 +97,33 @@ local function tags_lsp_to_vim(diagnostic, client_id)
+---@param offset_encoding string
+---@return vim.DiagnosticRelatedInformation?
+local function related_information_lsp_to_vim(related_informations, offset_encoding)
+  return vim.tbl_filter(function(item) return item ~= nil end,
+    --- @param related_information lsp.DiagnosticRelatedInformation
+    --- @return vim.DiagnosticRelatedInformation
+    vim.tbl_map(function(related_information)
+      local fname = vim.uri_to_fname(related_information.location.uri)
+      local bufnr = vim.fn.bufadd(fname)
+      if not bufnr then
+        return
+      end
+      local buf_lines = get_buf_lines(bufnr)
+      local start = related_information.location.range.start
+      local _end = related_information.location.range['end']
+      --- @type vim.DiagnosticRelatedInformation
+      return {
+        lnum = start.line,
+        col = line_byte_from_position(buf_lines, start.line, start.character, offset_encoding),
+        end_lnum = _end.line,
+        end_col = line_byte_from_position(buf_lines, _end.line, _end.character, offset_encoding),
+        message = related_information.message,
+      }
+    end, related_informations))
```
Chaining a map with a filter is a good use case for `vim.iter`. `vim.iter(related_informations):map()` gives you filter+map semantics combined (if the callback function returns `nil` then the item is filtered out).

Although, is the original filter for `item ~= nil` even necessary? Presumably `related_informations` is an array, so it should not have any `nil` items in it.

tom-anders replied:
The intention was to handle the `if not bufnr then return end` path - But thanks for the hint about using `vim.iter(..):map` (the rustacean in me would've expected this to be called `filter_map` though)

gpanders commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -97,6 +97,33 @@ local function tags_lsp_to_vim(diagnostic, client_id)
+---@return vim.DiagnosticRelatedInformation[]
+local function related_information_lsp_to_vim(related_informations, offset_encoding)
+  return vim.iter(related_informations):map(
+  --- @param related_information lsp.DiagnosticRelatedInformation
+  --- @return vim.DiagnosticRelatedInformation
```
```suggestion
    --- @param related_information lsp.DiagnosticRelatedInformation
    --- @return vim.DiagnosticRelatedInformation?
```

gpanders commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -97,6 +97,33 @@ local function tags_lsp_to_vim(diagnostic, client_id)
+  --- @return vim.DiagnosticRelatedInformation
+    function(related_information)
+      local fname = vim.uri_to_fname(related_information.location.uri)
+      local bufnr = vim.fn.bufadd(fname)
+      if not bufnr then
+        return
+      end
```
It's functionally equivalent in this case but imo is slightly clearer (we explicitly return either `nil` or a `vim.DiagnosticRelatedInformation` on all code paths).

```suggestion
      local bufnr = vim.fn.bufadd(fname)
      if not bufnr then
        return nil
      end
```

gpanders commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -97,6 +97,33 @@ local function tags_lsp_to_vim(diagnostic, client_id)
+        end_col = line_byte_from_position(buf_lines, _end.line, _end.character, offset_encoding),
+        message = related_information.message,
+      }
+    end)
```
You also need `:totable()` to collapse the iterator back into a table (analagous to `collect()` in Rust)

```suggestion
    end):totable()
```

tom-anders commented:
Thanks for the review! I extended the tests in `lsp/diagnostic_spec.lua` now.

wookayin commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -42,6 +45,29 @@ local M = {}
+--- Related message and location for a diagnostic.
+--- This can be used to point to code locations that cause or are related to
+--- a diagnostic, e.g when duplicating a symbol in a scope.
+--- @class vim.DiagnosticRelatedInformation
```
How about having a namescope, like `vim.diagnostic.RelatedInformation`?

tom-anders replied:
Good point, done

gpanders commented on runtime/doc/news.txt:
```diff
@@ -271,6 +271,10 @@ The following new APIs and features were added.
   • |vim.lsp.buf.definition()|, |vim.lsp.buf.declaration()|,
     |vim.lsp.buf.type_definition()|, and |vim.lsp.buf.implementation()| now
     support the `loclist` field of |vim.lsp.ListOpts|.
+  • Add `vim.Diagnostic.related_information` field and fill it with
+    `lsp.DiagnosticRelatedInformation[]`. Deprecate
```
```suggestion
  • Add `vim.Diagnostic.related_information` field and fill it with
    `vim.diagnostic.RelatedInformation[]`. Deprecate
```

tom-anders replied:
I intentionally wrote `lsp.DiagnosticRelatedInformation` here though, to highlight the new logic I added in `lsp/diagnostic.lua`

gpanders replied:
`lsp.DiagnosticRelatedInformation[]` does not link to anything in Nvim's help docs so I'm not sure it's valuable to have here in the user facing news. `vim.diagnostic.RelatedInformation` does have a help tag though. Maybe we keep this simple and just say:

```suggestion
• Add a new `related_information` field to |vim.Diagnostic| in favor of
  `vim.Diagnostic.user_data.lsp.relatedInformation` (now deprecated). See
  |vim.diagnostic.RelatedInformation|.
```

tom-anders replied:
sounds good, done

tom-anders commented:
(rebased to fix conflicts)

gpanders commented:
Needs another rebase.

mfussenegger reviewed (commented):
Looks like this is missing the reverse mapping from vim to lsp. (And I'm starting to think we could just keep the original payload in a _lsp field on the vim.Diagnostic table, to ensure we send the original to the server and avoid any conversion errors)

tom-anders commented:
> Looks like this is missing the reverse mapping from vim to lsp. (And I'm starting to think we could just keep the original payload in a _lsp field on the vim.Diagnostic table, to ensure we send the original to the server and avoid any conversion errors)

Thanks for the hint - We already have `vim.diagnostic.user_data.lsp` which stored a subset of the original LSP diagnostic, I now simply made it store the entire table instead. That should also take care of the conversion back to LSP (Since `diagnostic_vim_to_lsp` already merges the `user_data.lsp` table

mfussenegger commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -118,17 +150,15 @@ local function diagnostic_lsp_to_vim(diagnostics, bufnr, client_id)
-          relatedInformation = diagnostic.relatedInformation,
-          data = diagnostic.data,
-        },
+        lsp = diagnostic,
```
Can you also update `diagnostic_vim_to_lsp` to make use of this?

tom-anders replied:
See my top level comment: if I understand the code correctly, `diagnostic_vim_to_lsp` already makes use of this, since it already merges the contents of `user_data.lsp` into the returned table. (This is also why the unit tests currently fails, I fixed this now in the latest push)

mfussenegger commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -97,6 +97,38 @@ local function tags_lsp_to_vim(diagnostic, client_id)
+      --- @param related_information lsp.DiagnosticRelatedInformation
+      --- @return vim.diagnostic.RelatedInformation?
+      function(related_information)
+        local fname = vim.uri_to_fname(related_information.location.uri)
```
@gpanders what do you think of making the related_information property a iterable or table?

It would allow us to make this lazy, so that users not using the information wouldn't have to pay for the conversion costs here

justinmk commented on runtime/doc/diagnostic.txt:
```diff
@@ -597,6 +601,23 @@ Lua module: vim.diagnostic                                    *diagnostic-api*
       • {virt_text_win_col}?  (`integer`) See |nvim_buf_set_extmark()|.
       • {virt_text_hide}?     (`boolean`) See |nvim_buf_set_extmark()|.
 
+*vim.diagnostic.RelatedInformation*
```
any problem with using a less verbose name?
```suggestion
*vim.diagnostic.Related*
```

tom-anders commented:
Hi everyone, I shifted focus to other projects and currently don't have time to work on this PR anymore. If somebody wants to take over, feel free to do so
